src/Multicall3.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.11;
+
+/// @title Multicall3
+/// @notice Aggregate results from multiple read-only function calls
+/// @dev Multicall & Multicall2 backwards-compatible
+/// @author Michael Elliot <mike@makerdao.com>
+/// @author Joshua Levine <joshua@makerdao.com>
+/// @author Nick Johnson <arachnid@notdot.net>
+/// @author Andreas Bigger <andreas@nascent.xyz>
+
+contract Multicall3 {
+    struct Call {
+        address target;
+        bytes callData;
+    }
+
+    struct Call3 {
+        address target;
+        bytes callData;
+        bool requireSuccess;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    function aggregate(Call[] memory calls) public returns (uint256 blockNumber, bytes[] memory returnData) {
+        blockNumber = block.number;
+        returnData = new bytes[](calls.length);
+        for(uint256 i = 0; i < calls.length; i++) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(success, "Multicall aggregate: call failed");
+            returnData[i] = ret;
+        }
+    }
+
+    function tryAggregate(bool requireSuccess, Call[] memory calls) public returns (Result[] memory returnData) {
+        returnData = new Result[](calls.length);
+        for(uint256 i = 0; i < calls.length; i++) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+
+            if (requireSuccess) {
+                require(success, "Multicall2 aggregate: call failed");
+            }
+
+            returnData[i] = Result(success, ret);
+        }
+    }
+
+    function tryBlockAndAggregate(bool requireSuccess, Call[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    function blockAndAggregate(Call[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    function aggregate3(Call3[] memory calls) public returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = new Result[](calls.length);
+        for(uint256 i = 0; i < calls.length; i++) {
+            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
+            require(!calls[i].requireSuccess || success, "Multicall3 aggregate3: call failed");
+            returnData[i] = Result(success, ret);
+        }
```

Comment:
Two small things to reduce gas usage which increases max number of calls:
1. Saving off `uint256 length = calls.length` in memory will be cheaper than accessing `calls.length` each iteration, for large arrays
2. Change for loop to:

```suggestion
        for(uint256 i = 0; i < calls.length;) {
            (bool success, bytes memory ret) = calls[i].target.call(calls[i].callData);
            require(!calls[i].requireSuccess || success, "Multicall3 aggregate3: call failed");
            returnData[i] = Result(success, ret);
            unchecked { i++ }
        }
```

More info on both of these here: https://gist.github.com/hrkrshnn/ee8fabd532058307229d65dcd5836ddc

And actually, we can do this for all `for` loops in this file, not just the `aggregate3` loop